asdf/commands/edit.py

```diff
@@ -0,0 +1,618 @@
+        # Create a YAML tree to validate
+        # The YAML text must be converted to a stream.
+        tree = yamlutil.load_tree(io.BytesIO(yaml_content))
+        if tree is None:
+            print("Error: 'yamlutil.load_tree' failed to return a tree.")
+            sys.exist(1)
+
+        schema.validate(tree)  # Failure raises an exception.
+
+    return yaml_content
+
+
+def edit_func(fname, oname):
+    """
+    Creates a YAML file from an ASDF file.  The YAML file will contain only the
+    YAML from the ASDF file.  The YAML text will be written to a YAML text file
+    in the same, so from 'example.asdf' the file 'example.yaml' will be created.
+
+    Parameters
+    ----------
+    fname : str
+        Input ASDF file name
+    oname : str
+        Output YAML file name.
+    """
+    if not is_valid_asdf_path(fname):
+        return False
+
+    if not is_yaml_file(oname):
+        print("A YAML file is expected, with '.yaml' or '.yml' extension.")
+        sys.exit(1)
+
+    # Validate input file is an ASDF file.
+    fd, asdf_text = open_and_check_asdf_header(fname)
+
+    # Read and validate the YAML of an ASDF file.
+    yaml_text = read_and_validate_yaml(fd, fname, False)
+    fd.close()
+
+    # Write the YAML for the original ASDF file.
+    with open(oname, "wb") as ofd:
+        ofd.write(asdf_text)
+        ofd.write(yaml_text)
+
+    # Output message to user.
+    delim = "*" * 70
+    print(f"\n{delim}")
+    print("ASDF formatting and YAML schema validated.")
+    print(f"The text portion of '{fname}' is written to:")
+    print(f"    '{oname}'")
+    print(f"The file '{oname}' can be edited using your favorite text editor.")
+    print("The edited text can then be saved to the ASDF file of your choice")
+    print("using 'asdftool edit -s -f <edited text file> -o <ASDF file>.")
+    print(f"{delim}\n")
+
+    return
+
+
+def write_block_index(fd, index):
+    """
+    Write the block index to an ASDF file.
+
+    Parameters
+    ----------
+    fd : file descriptor
+    index : list
+        Integer location for each block.
+    """
+    global yaml_version
+    if len(index) < 1:
+        return
+
+    fd.write(constants.INDEX_HEADER)
+    fd.write(b"\n")
+
+    # If no YAML version found in edited YAML force it to 1.1
+    if yaml_version is None:
+        yaml_version = tuple([1, 1])
+    yaml.dump(
+        index,
+        Dumper=yamlutil._yaml_base_dumper,
+        stream=fd,
+        explicit_start=True,
+        explicit_end=True,
+        version=yaml_version,
+        allow_unicode=True,
+        encoding="utf-8",
+    )
+
+
+def find_first_block(fname):
+    """
+    Finds the location of the first binary block in an ASDF file.
+
+    Parameters
+    ----------
+    fname : str
+        Input ASDF file name.
+
+    Return
+    ------
+    int
+        Location, in bytes, of the first binary block.
+    """
+    with generic_io.get_file(fname, mode="r") as fd:
+        reader = fd.reader_until(
```

Comment:
Is it safe to seek from the beginning of the file to the start of the block area in this way?  All the bytes in `constants.BLOCK_MAGIC` are in the [YAML valid ranges](https://yaml.org/spec/1.1/current.html#id868518).  In asdf.py we only start looking for the first block after we've consumed the YAML document.

Reply:
I changed the search to first find the YAML ending delimiter, then find the first block after that delimiter.